### PR TITLE
Rule Manager: support resource level rule replacement in flow module

### PR DESCRIPTION
**Pull request by louyuting**
### Describe what this PR does / why we need it
Rule Manager: support resource level rule replacement in flow module

### Special notes for reviews

### CODE REVIEW

File: core/flow/rule_manager.go [hunk 1]
```diff
@@ -224,27 +230,106 @@ func onRuleUpdate(rules []*Rule) (err error) {
...
-	logRuleUpdate(resRulesMap)
+	logRuleUpdate(validResRulesMap)
 	return nil
 }
 
 // LoadRules loads the given flow rules to the rule manager, while all previous rules will be replaced.
 // the first returned value indicates whether do real load operation, if the rules is the same with previous rules, return false
 func LoadRules(rules []*Rule) (bool, error) {
-	// TODO: rethink the design
+	resRulesMap := make(map[string][]*Rule, 9)
```
Comment (luckyxiaoqiang) [inline core/flow/rule_manager.go:244]: `9` may be a little strange for people who aren't familiar with the implementation details. How about 16?

File: core/flow/rule_manager.go [hunk 2]
```diff
@@ -224,27 +230,106 @@ func onRuleUpdate(rules []*Rule) (err error) {
...
+	logging.Debug("[Flow onRuleUpdate] Time statistic(ns) for updating flow rule", "timeCost", util.CurrentTimeNano()-start)
+	logging.Info("[Flow] load resource level rules", "resource", res, "resRules", resRules)
+	return nil
+}
+
+// LoadRulesOfResource loads the given resource's flow rules to the rule manager, while all previous resource's rules will be replaced.
+// the first returned value indicates whether do real load operation, if the rules is the same with previous resource's rules, return false
+func LoadRulesOfResource(res string, rules []*Rule) (bool, error) {
+	if len(res) == 0 {
+		return false, errors.New("empty resource")
```
Comment (luckyxiaoqiang) [inline core/flow/rule_manager.go:310]: Zero length of rules means clear the resource's rules.

File: core/flow/rule_manager.go [hunk 3]
```diff
@@ -224,27 +230,106 @@ func onRuleUpdate(rules []*Rule) (err error) {
...
+	defer updateRuleMux.Unlock()
+	// clear resource rules
+	if len(rules) == 0 {
+		// clear resource's currentRules
+		delete(currentRules, res)
+		// clear tcMap
+		tcMux.Lock()
+		delete(tcMap, res)
+		tcMux.Unlock()
+		logging.Info("[Flow] clear resource level rules", "resource", res, "rules", rules)
```
Comment (luckyxiaoqiang) [inline core/flow/rule_manager.go:322]: `rules` is empty, so no need to print it.

File: core/flow/rule_manager.go [hunk 4]
```diff
@@ -224,27 +230,106 @@ func onRuleUpdate(rules []*Rule) (err error) {
...
+			if !ok {
+				err = fmt.Errorf("%v", r)
+			}
+		}
+	}()
+
+	validResRules := make([]*Rule, 0, len(resRules))
+	for _, rule := range resRules {
+		if err := IsValidRule(rule); err != nil {
+			logging.Warn("[Flow onRuleUpdate] Ignoring invalid flow rule", "rule", rule, "reason", err.Error())
```
Comment (luckyxiaoqiang) [inline core/flow/rule_manager.go:278]: [Flow onRuleUpdate] ==> [Flow onResourceRuleUpdate]

File: core/flow/rule_manager_test.go [hunk 5]
```diff
@@ -493,3 +499,138 @@ func TestIsValidRule(t *testing.T) {
...
+		Resource:               "abc2",
+		Threshold:              20,
+		TokenCalculateStrategy: Direct,
+		ControlBehavior:        Reject,
+	}
+
+	succ, err := LoadRules([]*Rule{&r11, &r12, &r21, &r22})
+	assert.True(t, succ && err == nil)
+
+	t.Run("Test_onResourceRuleUpdate-", func(t *testing.T) {
```
Comment (luckyxiaoqiang) [inline core/flow/rule_manager_test.go:582]: Does the `-` suffix have special usage?

File: core/flow/rule_manager.go (hunk 1 above)
Comment (louyuting) [reply core/flow/rule_manager.go:244]: ok

File: core/flow/rule_manager.go (hunk 4 above)
Comment (louyuting) [reply core/flow/rule_manager.go:278]: updated

File: core/flow/rule_manager.go (hunk 3 above)
Comment (louyuting) [reply core/flow/rule_manager.go:322]: updated

File: core/flow/rule_manager_test.go (hunk 5 above)
Comment (louyuting) [reply core/flow/rule_manager_test.go:582]: type fixed

File: core/flow/rule_manager_test.go [hunk 6]
```diff
@@ -493,3 +499,138 @@ func TestIsValidRule(t *testing.T) {
...
+		Resource:               "abc2",
+		Threshold:              20,
+		TokenCalculateStrategy: Direct,
+		ControlBehavior:        Reject,
+	}
+
+	succ, err := LoadRules([]*Rule{&r11, &r12, &r21, &r22})
+	assert.True(t, succ && err == nil)
+
+	t.Run("TestClearRulesOfResource-", func(t *testing.T) {
```
Comment (luckyxiaoqiang) [inline core/flow/rule_manager_test.go:627]: Typo?

Comment (louyuting) [reply core/flow/rule_manager_test.go:627]: Yes，I will fix it

**Comment by louyuting** [comment]
@liqiangz  Could you please review it?

File: core/flow/rule_manager.go [hunk 7]
```diff
@@ -186,32 +186,38 @@ func onRuleUpdate(rules []*Rule) (err error) {
...
 
-	resRulesMap := make(map[string][]*Rule, len(rules))
-	for _, rule := range rules {
-		if err := IsValidRule(rule); err != nil {
-			logging.Warn("[Flow onRuleUpdate] Ignoring invalid flow rule", "rule", rule, "reason", err.Error())
+	// ignore invalid rules
+	validResRulesMap := make(map[string][]*Rule, len(rawResRulesMap))
+	for res, rules := range rawResRulesMap {
+		if len(res) == 0 {
+			logging.Warn("[Flow onRuleUpdate] Ignoring empty resource", "rules", rules)
```
Comment (liqiangz) [inline core/flow/rule_manager.go:193]: IsValidRule() check empty Resource, Maybe this check is redundant?
```golang
func IsValidRule(rule *Rule) error {
	if rule == nil {
		return errors.New("nil Rule")
	}
	if rule.Resource == "" {
		return errors.New("empty Resource")
	}
```

Comment (louyuting) [reply core/flow/rule_manager.go:193]: Yes, I will remove here.

**Review by luckyxiaoqiang — DISMISSED** [review]

File: core/flow/rule_manager_test.go [hunk 8]
```diff
@@ -493,3 +499,138 @@ func TestIsValidRule(t *testing.T) {
...
+		Resource:               "abc2",
+		Threshold:              20,
+		TokenCalculateStrategy: Direct,
+		ControlBehavior:        Reject,
+	}
+
+	succ, err := LoadRules([]*Rule{r11, r12, r21, r22})
+	assert.True(t, succ && err == nil)
+
+	t.Run("LoadRulesOfResource-empty-resource", func(t *testing.T) {
```
Comment (sanxun0325) [inline core/flow/rule_manager_test.go:532]: test 命名目前都是代码中都是下划线或者驼峰。

File: core/flow/rule_manager.go [hunk 9]
```diff
@@ -224,27 +231,110 @@ func onRuleUpdate(rules []*Rule) (err error) {
...
+	tcMux.Lock()
+	if len(newResTcs) == 0 {
+		delete(tcMap, res)
+	} else {
+		tcMap[res] = newResTcs
+	}
+	tcMux.Unlock()
+	currentRules[res] = resRules
+	logging.Debug("[Flow onRuleUpdate] Time statistic(ns) for updating flow rule", "timeCost", util.CurrentTimeNano()-start)
+	logging.Info("[Flow] load resource level rules", "resource", res, "resRules", resRules)
```
Comment (liqiangz) [inline core/flow/rule_manager.go:307]: MayBe it should log validResRulesMap

File: core/flow/rule_manager.go [hunk 10]
```diff
@@ -224,27 +231,110 @@ func onRuleUpdate(rules []*Rule) (err error) {
...
+	isEqual := reflect.DeepEqual(currentRules, resRulesMap)
 	if isEqual {
 		logging.Info("[Flow] Load rules is the same with current rules, so ignore load operation.")
 		return false, nil
 	}
+	err := onRuleUpdate(resRulesMap)
+	return true, err
+}
+
+func onResourceRuleUpdate(res string, resRules []*Rule) (err error) {
```
Comment (liqiangz) [inline core/flow/rule_manager.go:265]: Maybe resRules should rename to rawResRules, in order to be consistent with onRuleUpdate method parameters

File: core/flow/rule_manager_test.go [hunk 11]
```diff
@@ -169,9 +171,7 @@ func TestGetRules(t *testing.T) {
 		assert.True(t, len(tcMap["abc2"]) == 1 && !tcMap["abc2"][0].boundStat.reuseResourceStat)
 		assert.True(t, reflect.DeepEqual(tcMap["abc2"][0].boundStat.readOnlyMetric, nopStat.readOnlyMetric))
 		assert.True(t, reflect.DeepEqual(tcMap["abc2"][0].boundStat.writeOnlyMetric, nopStat.writeOnlyMetric))
-		if err := ClearRules(); err != nil {
```
Comment (liqiangz) [inline core/flow/rule_manager_test.go:172]: If we remove ClearRules() there,   Maybe we should add ClearRules() test case,

File: core/flow/rule_manager.go (hunk 9 above)
Comment (louyuting) [reply core/flow/rule_manager.go:307]: make sense. invalid rule had been logged in previous logic.

File: core/flow/rule_manager.go (hunk 10 above)
Comment (louyuting) [reply core/flow/rule_manager.go:265]: updated

File: core/flow/rule_manager_test.go (hunk 11 above)
Comment (louyuting) [reply core/flow/rule_manager_test.go:172]: `clearData()` do it~

File: core/flow/rule_manager_test.go (hunk 8 above)
Comment (louyuting) [reply core/flow/rule_manager_test.go:532]: updated

**Review by liqiangz — APPROVED** [review]
LGTM

**Review by luckyxiaoqiang — APPROVED** [review]

**Pull request merged**